pyro/contrib/gp/models/gpr.py

```diff
@@ -0,0 +1,73 @@
+from __future__ import absolute_import, division, print_function
+
+from torch.autograd import Variable
+import torch.nn as nn
+
+import pyro
+import pyro.distributions as dist
+
+
+class GPRegression(nn.Module):
+    """
+    Gaussian Process regression module.
+
+    :param torch.autograd.Variable X: A tensor of inputs.
+    :param torch.autograd.Variable y: A tensor of outputs for training.
+    :param pyro.gp.kernels.Kernel kernel: A Pyro kernel object.
+    :param torch.Tensor noise: An optional noise tensor.
+    :param dict priors: A mapping from kernel parameter's names to priors.
+    """
+    def __init__(self, X, y, kernel, noise=None, priors=None):
+        super(GPRegression, self).__init__()
+        self.X = X
+        self.y = y
+        self.input_dim = X.size(0)
+        self.kernel = kernel
+        # TODO: define noise as a nn.Module, so we can train/set prior to it
+        if noise is None:
+            self.noise = Variable(X.data.new([1]))
+        else:
+            self.noise = Variable(noise)
+        self.priors = priors
+        if priors is None:
+            self.priors = {}
+
+    def model(self):
+        kernel_fn = pyro.random_module(self.kernel.name, self.kernel, self.priors)
+        kernel = kernel_fn()
+        K = kernel(self.X) + self.noise.repeat(self.input_dim).diag()
+        zero_loc = Variable(K.data.new([0]).expand(self.input_dim))
+        pyro.sample("f", dist.MultivariateNormal(zero_loc, K), obs=self.y)
+
+    def guide(self):
```

Comment:
What is the purpose of the guide in this context? It seems like you are doing inference in forward(), so what is the point of the guide?

Reply:
Originally, I want to put some constraints on parameters but it needs to write a wrapper of `nn.Parameter` (support transform methods). Then I found that setting priors and using guide for MAP inference might be a simpler idea.  Do you find a better way of using the guide?